# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/nsAccessibilityService.cpp
# Commit: 2ba84268f8ac
# Full Hash: 2ba84268f8ac2baf36477e62bc03ade97966d865
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2025-09-30 17:19 +0000
# Regressor Bug: 1987945
# File Overlap Count: 1
# Overlapping Files: accessible/tests/browser/relations/browser_anchor_positioning.js
# Description:
#   Bug 1987945 - P3: Support details relation in anchored targets in remote accessibles. r=Jamie,layout-anchor-positioning-reviewers,layout-reviewers,emilio,firefox-style-system-reviewers
# ==============================================================================

diff -r 7aaaa4514127 -r 2ba84268f8ac accessible/base/nsAccessibilityService.cpp
--- a/accessible/base/nsAccessibilityService.cpp	Tue Sep 30 17:19:31 2025 +0000
+++ b/accessible/base/nsAccessibilityService.cpp	Tue Sep 30 17:19:31 2025 +0000
@@ -674,6 +674,48 @@
   }
 }
 
+void nsAccessibilityService::NotifyAnchorPositionedRemoved(
+    mozilla::PresShell* aPresShell, nsIFrame* aFrame) {
+  DocAccessible* document = aPresShell->GetDocAccessible();
+  if (!document) {
+    return;
+  }
+
+  nsIFrame* anchorFrame =
+      nsCoreUtils::GetAnchorForPositionedFrame(aPresShell, aFrame);
+  if (!anchorFrame) {
+    return;
+  }
+
+  if (LocalAccessible* anchorAcc =
+          document->GetAccessible(anchorFrame->GetContent())) {
+    document->QueueCacheUpdate(anchorAcc, CacheDomain::Relations);
+  }
+}
+
+void nsAccessibilityService::NotifyAnchorRemoved(mozilla::PresShell* aPresShell,
+                                                 nsIFrame* aFrame) {
+  DocAccessible* document = aPresShell->GetDocAccessible();
+  if (!document) {
+    return;
+  }
+
+  nsIFrame* positionedFrame =
+      nsCoreUtils::GetPositionedFrameForAnchor(aPresShell, aFrame);
+  if (!positionedFrame) {
+    return;
+  }
+
+  if (LocalAccessible* positionedAcc =
+          document->GetAccessible(positionedFrame->GetContent())) {
+    // If the anchor was removed, its positioned element may now have a 1:1
+    // relation with another anchor, and they would get a description a11y
+    // relation. So we need to go one level deeper here and refresh the cache of
+    // any potential anchors that remain on the positioned element.
+    document->RefreshAnchorRelationCacheForTarget(positionedAcc);
+  }
+}
+
 void nsAccessibilityService::NotifyAttrElementWillChange(
     mozilla::dom::Element* aElement, nsAtom* aAttr) {
   mozilla::dom::Document* doc = aElement->OwnerDoc();